purchases/src/main/kotlin/com/revenuecat/purchases/common/events/EventsManager.kt

```diff
@@ -0,0 +1,187 @@
+
+internal class EventsManager(
+    // legacy events store for paywalls
+    private val legacyEventsFileHelper: EventsFileHelper<PaywallStoredEvent>,
+    private val fileHelper: EventsFileHelper<BackendEvent>,
+    private val identityManager: IdentityManager,
+    private val eventsDispatcher: Dispatcher,
+    private val postEvents: (
+        EventRequest,
+        () -> Unit,
+        (error: PurchasesError, shouldMarkAsSynced: Boolean) -> Unit,
+    ) -> Unit,
+    private val flushCount: Int = 50,
+) {
+
+    companion object {
+        const val PAYWALL_EVENTS_FILE_PATH = "RevenueCat/paywall_event_store/paywall_event_store.jsonl"
+        const val EVENTS_FILE_PATH_NEW = "RevenueCat/event_store/event_store.jsonl"
+
+        internal val json = Json {
+            serializersModule = SerializersModule {
+                polymorphic(BackendEvent::class) {
+                    subclass(BackendEvent.CustomerCenter::class, BackendEvent.CustomerCenter.serializer())
+                    subclass(BackendEvent.Paywalls::class, BackendEvent.Paywalls.serializer())
+                }
+            }
+        }
+
+        fun backendEvents(fileHelper: FileHelper): EventsFileHelper<BackendEvent> {
+            return EventsFileHelper(
+                fileHelper,
+                EVENTS_FILE_PATH_NEW,
+                { event -> json.encodeToString(BackendEvent.serializer(), event) },
+                { jsonString -> json.decodeFromString(BackendEvent.serializer(), jsonString) },
+            )
+        }
+
+        fun paywalls(fileHelper: FileHelper): EventsFileHelper<PaywallStoredEvent> {
+            return EventsFileHelper(
+                fileHelper,
+                EventsManager.PAYWALL_EVENTS_FILE_PATH,
+                PaywallStoredEvent::toString,
+                PaywallStoredEvent::fromString,
+            )
+        }
+    }
+
+    @get:Synchronized
+    @set:Synchronized
+    private var flushInProgress = false
+
+    @get:Synchronized
+    @set:Synchronized
+    private var legacyFlushDone = false
+
+    @OptIn(ExperimentalPreviewRevenueCatPurchasesAPI::class)
+    @Synchronized
+    fun track(event: FeatureEvent) {
+        enqueue {
+            debugLog("Tracking event: $event")
+
+            val backendEvent = when {
+                event.paywallEvent != null -> event.paywallEvent!!.toBackendEvent(identityManager.currentAppUserID)
+                else -> null
+            }
+
+            if (backendEvent != null) {
+                fileHelper.appendEvent(backendEvent)
+            } else {
+                debugLog("Backend event not implemented for: $event")
+            }
+        }
+    }
+
+    @Synchronized
+    fun flushEvents() {
+        enqueue {
+            if (flushInProgress) {
+                debugLog("Flush already in progress.")
+                return@enqueue
+            }
+            flushInProgress = true
+
+            if (!legacyFlushDone) {
+                flushLegacyEvents()
```

Comment:
this is being enqueued, so I would not call `legacyFlushDone` done, since it's actually not guaranteed it's been completed. Maybe `legacyFlushTriggered`?

Reply:
actually since it's enqueued... there's a race condition if `flushEvents` executes again before flushLegacyEvents finishes? A way to solve it is removing the enqueuing in `flushLegacyEvents`. Or actually... shouldn't it go before `flushLegacyEvents` and not after?

Reply:
I'll put it before since this is just a fire-and-forget thing.